src/main/java/bigqueryestatespring/services/EstatesService.java

```diff
@@ -0,0 +1,291 @@
+        }
+    }
+
+    /**
+     * Constructor with default operation type (AVG)
+     *
+     * @param columnNames list of columns to create tree (last element will be processed in aggregate function)
+     */
+    public EstatesService(List<String> columnNames) {
+        this.columnNames = columnNames;
+    }
+
+    /**
+     * Gets bigQuery instance with credentials for estates project
+     *
+     * @return bigQuery instance
+     */
+    private BigQuery getBigQueryInstance() {
+        try {
+            return BigQueryOptions.newBuilder()
+                    .setCredentials(
+                            ServiceAccountCredentials
+                                    .fromStream(new FileInputStream(PATH_TO_CREDENTIALS))
+                    ).build().getService();
+        } catch (FileNotFoundException ex) {
+            ex.printStackTrace();
+            throw new RuntimeException(CREDENTIALS_FILE_NOT_FOUND);
+        } catch (IOException ex) {
+            ex.printStackTrace();
+            throw new RuntimeException(EXCEPTION_WHILE_READING_CREDENTIALS);
+        }
+    }
+
+    /**
+     * Gets QueryJobConfiguration which contains sql query as
+     * SELECT columnNames(except last), operationType(columnNames(last))
+     * FROM *** WHERE surface >= bottom and surface <= top
+     * GROUP BY columnNames(except last)
+     *
+     * @param bottom low border for space
+     * @param top high border for space
+     * @return QueryJobConfiguration
+     */
+    private QueryJobConfiguration getQueryJobConfiguration(int bottom, int top) {
+        StringBuilder queryStrBuilder = new StringBuilder("SELECT ");
+        if (columnNames != null && !columnNames.isEmpty()) {
+            for (String columnName : columnNames) {
+                queryStrBuilder.append(columnName).append(", ");
+            }
+        }
+        queryStrBuilder.append(operationType.getStringValue()).append('(')
+                .append(aggregateColumn).append(')')
+                .append(" AS ").append(alias).append(' ')
+                .append("FROM `properati-data-public.properties_ar.properties_rent_201501` ")
+                .append("WHERE (surface_covered_in_m2 >= ")
+                .append(bottom)
+                .append(" AND surface_covered_in_m2 <= ")
+                .append(top)
+                .append(") ");
+
+        if (columnNames != null && !columnNames.isEmpty()) {
+            queryStrBuilder.append("GROUP BY ").append(columnNames.get(0));
+            for (String columnName : columnNames.subList(1, columnNames.size())) {
+                queryStrBuilder.append(", ").append(columnName);
+            }
+        }
+
+        return QueryJobConfiguration
+                    .newBuilder(queryStrBuilder.toString())
+                    .setUseLegacySql(false)
+                    .build();
+    }
+
+    private TableResult getTableResultOfEstates(int bottom, int top) {
```

Comment:
Big Query has some limitations of concurrent jobs running.
Can you use thread pool and use future here?